File: main/WiFiManager.swift
public enum WiFiManagerError: Swift.Error {
    case initializationFailed
    case apStartFailed
    case connectionFailed
    case invalidIPAddress
    case httpServerError
    case deinitializationFailed
}

public struct WiFiCredentials {
    let ssid: String
    let password: String

    public init(ssid: String, password: String = "") {
        self.ssid = ssid
        self.password = password
    }
}

public struct APConfiguration {
    let credentials: WiFiCredentials
    let channel: UInt8
    let maxConnections: UInt8

    public init(
        credentials: WiFiCredentials,
        channel: UInt8 = 1,
        maxConnections: UInt8 = 4
    ) {
        self.credentials = credentials
        self.channel = channel
        self.maxConnections = maxConnections
    }
}

public class ClosureWrapper {
    let closure: () -> Void
    
    init(_ closure: @escaping () -> Void) {
        self.closure = closure
    }
}

public final class WiFiManager {
    private var isInitialized = false
    private var handle: httpd_handle_t? = nil
    var handler: () -> Void = {}

    public init() {}

    public func startHTTPServer(test: @escaping () -> Void) throws(WiFiManagerError) {
        if !isInitialized {
            try initialize()
        }

        var config = DEFAULT_SERVER_CONFIG
        var localServer: httpd_handle_t?

        guard httpd_start(&localServer, &config) == ESP_OK,
            let server = localServer
        else {
            throw WiFiManagerError.httpServerError
        }

        self.handle = server

        // Store the test closure in an unmanaged pointer

        let commandUri = "/led"
        let closureWrapper = ClosureWrapper(test)
        let userContext = Unmanaged.passRetained(closureWrapper).toOpaque()
        // Pass a function pointer compatible closure
        var closure: @convention(c) (UnsafeMutablePointer<httpd_req_t>?) -> esp_err_t = { req in
            guard let req = req else { return ESP_FAIL }
            print("started")
            
            // Retrieve the closure from the unmanaged pointer            
            let bufferSize = 100
            var content = [CChar](repeating: 0, count: bufferSize)
            let recvSize = min(Int(req.pointee.content_len), bufferSize - 1)

            print("received!")

            guard httpd_req_recv(req, &content, recvSize) > 0 else {
                return ESP_FAIL
            }

            print(String(cString: content))

            let response = "Command received\n"

            guard let userContext = req.pointee.user_ctx else { return ESP_FAIL }
            let closureWrapper = Unmanaged<ClosureWrapper>.fromOpaque(userContext).takeUnretainedValue()
            let testClosure = closureWrapper.closure
            testClosure() // Call the closure

            response.withCString { cString in
                httpd_resp_send(req, cString, Int(HTTPD_RESP_USE_STRLEN))
            }

            return ESP_OK
        }

        var uriHandler = httpd_uri_t()
        uriHandler = httpd_uri_t(
            uri: commandUri,
            method: HTTP_POST,
            handler: closure,
            user_ctx: userContext
        )

        guard httpd_register_uri_handler(server, &uriHandler) == ESP_OK else {
            httpd_stop(server)
            Unmanaged<ClosureWrapper>.fromOpaque(userContext).release()
            throw WiFiManagerError.httpServerError
        }
    }

    public func initialize() throws(WiFiManagerError) {
        guard !isInitialized else { return }

        let result = wifi_manager_init()
        guard result == ESP_OK else {
            throw WiFiManagerError.initializationFailed
        }
        isInitialized = true
    }

    public func startAccessPoint(_ config: APConfiguration) throws(WiFiManagerError) {
        guard isInitialized else {
            try initialize()

            return
        }

        let result = config.credentials.ssid.withCString { ssid in
            config.credentials.password.withCString { password in
                wifi_manager_start_ap(
                    ssid,
                    password,
                    config.channel,
                    config.maxConnections
                )
            }
        }

        guard result == ESP_OK else {
            throw WiFiManagerError.apStartFailed
        }
    }

    public func connectToNetwork(
        _ credentials: WiFiCredentials,
        timeoutMs: UInt32 = 10000
    ) throws(WiFiManagerError) -> Bool {
        if !isInitialized {
            try initialize()
        }

        var connected = false
        let result = credentials.ssid.withCString { ssid in
            credentials.password.withCString { password in
                wifi_manager_connect_sta(
                    ssid,
                    password,
                    timeoutMs,
                    &connected
                )
            }
        }

        guard result == ESP_OK else {
            throw WiFiManagerError.connectionFailed
        }

        return connected
    }

    public func stopHTTPServer() throws(WiFiManagerError) {
        guard isInitialized else { return }

        let result = wifi_manager_stop_http_server()
        guard result == ESP_OK else {
            throw WiFiManagerError.httpServerError
        }
    }

    public func getAccessPointIPAddress() throws(WiFiManagerError) -> String {
        let bufferSize = 16  // Maximum IPv4 string length
        var buffer = [CChar](repeating: 0, count: bufferSize)

        let result = wifi_manager_get_ap_ip(&buffer, bufferSize)
        guard result == ESP_OK else {
            throw WiFiManagerError.invalidIPAddress
        }

        return String(cString: buffer)
    }

    public func getStationIPAddress() throws(WiFiManagerError) -> String {
        guard let netif = esp_netif_get_handle_from_ifkey("WIFI_STA_DEF") else {
            throw WiFiManagerError.invalidIPAddress
        }

        var ipInfo = esp_netif_ip_info_t()
        guard esp_netif_get_ip_info(netif, &ipInfo) == ESP_OK else {
            throw WiFiManagerError.invalidIPAddress
        }

        let ip = ipInfo.ip
        let addr: UInt32 = ipInfo.ip.addr
        let byte1: UInt8 = UInt8(addr & 0xFF)
        let byte2: UInt8 = UInt8((addr >> 8) & 0xFF)
        let byte3: UInt8 = UInt8((addr >> 16) & 0xFF)
        let byte4: UInt8 = UInt8((addr >> 24) & 0xFF)

        return "\(byte1).\(byte2).\(byte3).\(byte4)"
    }

    public func deinitialize() throws(WiFiManagerError) {
        guard isInitialized else { return }

        let result = wifi_manager_deinit()
        guard result == ESP_OK else {
            throw WiFiManagerError.deinitializationFailed
        }
        isInitialized = false
    }

    deinit {
        try? deinitialize()
    }
}

private let ESP_OK: esp_err_t = 0

File: main/src/wifi_manager.c
#include "esp_wifi.h"
#include "esp_event.h"
#include "esp_log.h"
#include "nvs_flash.h"
#include <string.h>
#include "freertos/event_groups.h"
#include "esp_netif.h"
#include "esp_http_server.h"

static const char *TAG = "wifi_manager";
static EventGroupHandle_t s_wifi_event_group;
static esp_netif_t *esp_netif_ap = NULL;
static esp_netif_t *esp_netif_sta = NULL;
static httpd_handle_t server = NULL;

#define WIFI_CONNECTED_BIT BIT0
#define WIFI_FAIL_BIT      BIT1

static void wifi_event_handler(void *arg, esp_event_base_t event_base,
                             int32_t event_id, void *event_data)
{
    if (event_base == WIFI_EVENT) {
        switch (event_id) {
            case WIFI_EVENT_STA_START:
                esp_wifi_connect();
                break;
            case WIFI_EVENT_STA_DISCONNECTED:
                xEventGroupSetBits(s_wifi_event_group, WIFI_FAIL_BIT);
                break;
            case WIFI_EVENT_AP_STACONNECTED:
                ESP_LOGI(TAG, "Station connected to AP");
                break;
            case WIFI_EVENT_AP_STADISCONNECTED:
                ESP_LOGI(TAG, "Station disconnected from AP");
                break;
        }
    } else if (event_base == IP_EVENT && event_id == IP_EVENT_STA_GOT_IP) {
        xEventGroupSetBits(s_wifi_event_group, WIFI_CONNECTED_BIT);
    }
}

static esp_err_t handle_post_request(httpd_req_t *req)
{
    char content[100];
    size_t recv_size = MIN(req->content_len, sizeof(content)-1);

    int ret = httpd_req_recv(req, content, recv_size);
    if (ret <= 0) {
        return ESP_FAIL;
    }
    content[ret] = '\0';

    ESP_LOGI(TAG, "Received POST data: %s", content);

    const char resp[] = "Command received\n";
    httpd_resp_send(req, resp, HTTPD_RESP_USE_STRLEN);
    return ESP_OK;
}

esp_err_t wifi_manager_init(void)
{
    esp_err_t ret = nvs_flash_init();
    if (ret == ESP_ERR_NVS_NO_FREE_PAGES || ret == ESP_ERR_NVS_NEW_VERSION_FOUND) {
        ESP_ERROR_CHECK(nvs_flash_erase());
        ret = nvs_flash_init();
    }
    ESP_ERROR_CHECK(ret);

    ESP_ERROR_CHECK(esp_netif_init());
    ESP_ERROR_CHECK(esp_event_loop_create_default());
    
    s_wifi_event_group = xEventGroupCreate();
    
    wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
    ESP_ERROR_CHECK(esp_wifi_init(&cfg));
    
    ESP_ERROR_CHECK(esp_event_handler_instance_register(WIFI_EVENT,
                    ESP_EVENT_ANY_ID,
                    &wifi_event_handler,
                    NULL,
                    NULL));
    ESP_ERROR_CHECK(esp_event_handler_instance_register(IP_EVENT,
                    IP_EVENT_STA_GOT_IP,
                    &wifi_event_handler,
                    NULL,
                    NULL));
                    
    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_APSTA));
    
    return ESP_OK;
}

esp_err_t wifi_manager_start_ap(const char* ap_ssid, 
                               const char* ap_password,
                               uint8_t ap_channel,
                               uint8_t max_connections)
{
    esp_netif_ap = esp_netif_create_default_wifi_ap();
    
    wifi_config_t wifi_ap_config = {
        .ap = {
            .channel = ap_channel,
            .max_connection = max_connections,
            .authmode = WIFI_AUTH_WPA2_PSK,
            .pmf_cfg = {
                .required = false,
            },
        },
    };
    
    strncpy((char*)wifi_ap_config.ap.ssid, ap_ssid, sizeof(wifi_ap_config.ap.ssid));
    strncpy((char*)wifi_ap_config.ap.password, ap_password, sizeof(wifi_ap_config.ap.password));
    
    if (strlen(ap_password) == 0) {
        wifi_ap_config.ap.authmode = WIFI_AUTH_OPEN;
    }
    
    ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_AP, &wifi_ap_config));
    
    return ESP_OK;
}

esp_err_t wifi_manager_connect_sta(const char* ssid,
                                  const char* password,
                                  uint32_t timeout_ms,
                                  bool* connected)
{
    *connected = false;
    esp_netif_sta = esp_netif_create_default_wifi_sta();
    
    wifi_config_t wifi_sta_config = {
        .sta = {
            .scan_method = WIFI_ALL_CHANNEL_SCAN,
            .failure_retry_cnt = 5,
            .threshold.authmode = WIFI_AUTH_WPA2_PSK,
            .sae_pwe_h2e = WPA3_SAE_PWE_BOTH,
        },
    };
    
    strncpy((char*)wifi_sta_config.sta.ssid, ssid, sizeof(wifi_sta_config.sta.ssid));
    strncpy((char*)wifi_sta_config.sta.password, password, sizeof(wifi_sta_config.sta.password));
    
    ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_STA, &wifi_sta_config));
    ESP_ERROR_CHECK(esp_wifi_start());
    
    EventBits_t bits = xEventGroupWaitBits(s_wifi_event_group,
                                          WIFI_CONNECTED_BIT | WIFI_FAIL_BIT,
                                          pdTRUE,
                                          pdFALSE,
                                          pdMS_TO_TICKS(timeout_ms));
    
    if (bits & WIFI_CONNECTED_BIT) {
        ESP_LOGI(TAG, "Connected to AP");
        *connected = true;
        esp_netif_set_default_netif(esp_netif_sta);
        if (esp_netif_napt_enable(esp_netif_ap) != ESP_OK) {
            ESP_LOGE(TAG, "NAPT not enabled");
        }
    } else {
        ESP_LOGI(TAG, "Failed to connect to AP");
    }
    
    return ESP_OK;
}

esp_err_t wifi_manager_deinit(void)
{
    ESP_ERROR_CHECK(esp_wifi_stop());
    ESP_ERROR_CHECK(esp_wifi_deinit());
    if (s_wifi_event_group) {
        vEventGroupDelete(s_wifi_event_group);
        s_wifi_event_group = NULL;
    }
    return ESP_OK;
}

esp_err_t wifi_manager_get_ap_ip(char* ip_addr, size_t buffer_size) {
    esp_netif_ip_info_t ip_info;
    esp_netif_get_ip_info(esp_netif_ap, &ip_info);
    snprintf(ip_addr, buffer_size, IPSTR, IP2STR(&ip_info.ip));
    return ESP_OK;
}

const httpd_config_t DEFAULT_SERVER_CONFIG = HTTPD_DEFAULT_CONFIG();

esp_err_t wifi_manager_stop_http_server(void)
{
    if (server) {
        httpd_stop(server);
        server = NULL;
    }
    return ESP_OK;
}

esp_err_t wifi_manager_get_sta_ip(char* ip_addr, size_t buffer_size) {
    esp_netif_t* netif = esp_netif_get_handle_from_ifkey("WIFI_STA_DEF");
    if (netif == NULL) {
        return ESP_FAIL;
    }
    
    esp_netif_ip_info_t ip_info;
    esp_err_t ret = esp_netif_get_ip_info(netif, &ip_info);
    
    if (ret != ESP_OK) {
        return ret;
    }
    
    // Convert IP address to string
    snprintf(ip_addr, buffer_size, IPSTR, IP2STR(&ip_info.ip));
    return ESP_OK;
} 